src/runtime/tracing.cpp

```diff
@@ -10,7 +10,105 @@ typedef int32_t (*trace_fn)(void *, const halide_trace_event_t *);
 
 namespace Halide { namespace Runtime { namespace Internal {
 
-WEAK int halide_trace_file = 0;
+// A spinlock that allows for shared and exclusive access. It's
+// equivalent to a reader-writer lock, but in my case the "readers"
+// will actually be writing simultaneously to the trace buffer, so
+// that's a bad name. We use the __sync primitives used elsewhere in
+// the runtime for atomic work. They are well supported by clang.
+class SharedExclusiveSpinLock {
+    uint32_t lock;
+    const static uint32_t exclusive_held_mask = 0x80000000;
+    const static uint32_t exclusive_waiting_mask = 0x40000000;
+    const static uint32_t shared_mask = 0x3fffffff;
+
+public:
+    void acquire_shared() {
+        while (1) {
+            uint32_t x = lock & shared_mask;
+            if (__sync_bool_compare_and_swap(&lock, x, x + 1)) {
+                return;
+            }
+        }
+    }
+
+    void release_shared() {
+        __sync_fetch_and_sub(&lock, 1);
+    }
+
+    void acquire_exclusive() {
+        while (1) {
+            __sync_fetch_and_or(&lock, exclusive_waiting_mask);
+            if (__sync_bool_compare_and_swap(&lock, exclusive_waiting_mask, exclusive_held_mask)) {
+                return;
+            }
+        }
+    }
+
+    void release_exclusive() {
+        __sync_fetch_and_xor(&lock, exclusive_held_mask);
+    }
+
+    SharedExclusiveSpinLock() : lock(0) {}
+};
+
+class TraceBuffer {
+    SharedExclusiveSpinLock lock;
+    uint32_t cursor;
+    uint8_t buf[1024*1024];
+
+public:
+
+    // Attempt to atomically acquire space in the buffer to write a
+    // packet. Returns NULL if the buffer was full.
+    halide_trace_packet_t *try_acquire_packet(uint32_t size) {
+        lock.acquire_shared();
+        uint32_t my_cursor = __sync_fetch_and_add(&cursor, size);
+        if (my_cursor + size > sizeof(buf)) {
+            __sync_fetch_and_sub(&cursor, size);
+            lock.release_shared();
+            return NULL;
+        } else {
+            return (halide_trace_packet_t *)(buf + my_cursor);
+        }
+    }
+
+    // Wait for all writers to finish with their packets, stall any
+    // new writers, and flush the buffer to the fd.
+    void flush(void *user_context, int fd) {
+        lock.acquire_exclusive();
+        bool success = true;
+        if (cursor) {
+            success = (cursor == (uint32_t)write(fd, buf, cursor));
+            cursor = 0;
+        }
+        lock.release_exclusive();
+        halide_assert(user_context, success && "Could not write to trace file");
+    }
+
+    // Acquire and return a packet's worth of space in the trace
```

Comment:
Should document this routine returns with the lock held in shared mode and that release_packet must be called to balance this. Might be worth documenting that the region of the buffer returned is protected from other threads because the cursor cannot wrap around without transitioning through exclusive state, which cannot happen while the lock is held in shared mode. Thus no other thread can get a region overlapping this one.